crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1416,10 +1416,31 @@ impl<'db> ClassType<'db> {
         let DefinitionKind::Class(class_stmt_node) = definition.kind(db) else {
             panic!("Class type definition must have DefinitionKind::Class");
         };
-        class_stmt_node
-            .bases()
-            .iter()
-            .map(move |base_expr| definition_expression_ty(db, definition, base_expr))
+
+        let has_type_params = class_stmt_node.type_params.is_some();
+
+        let type_scope_info: Option<(ScopeId<'db>, &TypeInference<'db>)> = if has_type_params {
+            let file = definition.file(db);
+            let index = semantic_index(db, file);
+            // we need to use the type param'd scope
+            let type_param_scope = index
+                .node_scope(NodeWithScopeRef::ClassTypeParameters(class_stmt_node))
+                .to_scope_id(db, file);
+            Some((type_param_scope, infer_scope_types(db, type_param_scope)))
+        } else {
+            None
+        };
+
+        let mapper = move |base_expr: &ast::Expr| {
+            if has_type_params {
+                // Safety: we calculated the inference if has_type_params
+                let (type_param_scope, inferences) = type_scope_info.unwrap();
```

Comment:
Is there a reason to use `has_type_params` and then `type_scope_info.unwrap()` rather than just `if let Some(...) = type_scope_info {` ? They rely on the same invariants, but the latter seems simpler.

Reply:
Hmm... your suggestion is cleaner and this code isn't big, but I dislike the hiding of the "the branching is due to type parameter-ness".

I believe my concerns can be resolved with a more well thought out variable name for the scope info. Will ruminate on it overnight

Reply:
Changed this up to `bases_scope_info`. It's still an `Option` (instead of doing something like checking if `bases_scope` is equal to some other scope to determine specialized-ness), but good enough for me to feel comfortable with the cleaner `if let` solution.